core/indigo-core/molecule/molecule_cdxml_loader.h

```diff
@@ -0,0 +1,204 @@
+        }
+
+    private:
+        int val;
+    };
+
+
+    struct _ExtConnection
+    {
+        int bond_id;
+        int point_id;
+        int atom_idx;
+    };
+
+    struct CdxmlNode
+    {
+        CdxmlNode() : element(ELEM_C), type( kCDXNodeType_Element ) // Carbon by default
+        {
+        }
+        AutoInt id;
+        std::string label;
+        AutoInt element;
+        Vec3f pos;
+        int type;
+        AutoInt isotope;
+        AutoInt charge;
+        AutoInt radical;
+        AutoInt valence;
+        AutoInt hydrogens;
+        AutoInt stereo;
+        AutoInt enchanced_stereo;
+        AutoInt stereo_group;
+        AutoInt index;
+        bool is_not_list;
+        std::vector<AutoInt> element_list;
+        std::unordered_map<int, int> bond_id_to_connection_idx;
+        std::unordered_map<int, int> node_id_to_connection_idx;
+        std::vector<_ExtConnection> connections;
+    };
+
+    struct CdxmlBond
+    {
+        CdxmlBond() : order(1)
+        {
+        }
+        AutoInt id;
+        std::pair<AutoInt, AutoInt> be;
+        AutoInt order;
+        AutoInt stereo;
+        AutoInt dir;
+        bool swap_bond;
+    };
+
+    struct CdxmlBracket
+    {
+        CdxmlBracket() : repeat_pattern(RepeatingUnit::HEAD_TO_TAIL)
+        {
+        }
+        std::vector<AutoInt> bracketed_list;
+        int usage;
+        AutoInt repeat_count;
+        int repeat_pattern;
+        std::string sru_label;
+    };
+
+
+    inline std::vector<std::string> split(const std::string& str, char delim)
+    {
+        std::vector<std::string> strings;
+        size_t start;
+        size_t end = 0;
+        while ((start = str.find_first_not_of(delim, end)) != std::string::npos)
+        {
+            end = str.find(delim, start);
+            strings.push_back(str.substr(start, end - start));
+        }
+        return strings;
+    }
+
+    class MoleculeCdxmlLoader
+    {
+    public:
+        DECL_ERROR;
+
+        MoleculeCdxmlLoader(Scanner& scanner);
+
+        void loadMolecule(BaseMolecule& mol);
+
+        StereocentersOptions stereochemistry_options;
+        bool ignore_bad_valence;
+        bool _has_bounding_box;
+        Rect2f _cdxml_bbox;
+        AutoInt _cdxml_bond_length;
+        std::vector<CdxmlNode> _nodes;
+        std::vector<CdxmlBond> _bonds;
+        std::vector<CdxmlBracket> _brackets;
+
+    protected:
+        Scanner* _scanner;
+        const TiXmlNode* _fragment;
+        void _parseCDXMLAttributes(TiXmlAttribute* pAttr);
+        void _parseNode(CdxmlNode& node, TiXmlElement* pElem);
+        void _addNode(CdxmlNode& node);
+
+        void _parseBond(CdxmlBond& bond, TiXmlAttribute* pAttr);
+        void _addBond(CdxmlBond& node);
+
+        void _parseBracket(CdxmlBracket& bracket, TiXmlAttribute* pAttr);
+
+        void _applyDispatcher(TiXmlAttribute* pAttr, const std::unordered_map<std::string, std::function<void(std::string&)>>& dispatcher);
```

Comment:
Such a wordy type (`std::unordered_map<std::string, std::function<void(std::string&)>>& dispatcher`) is worth using or typedef.